### PR TITLE
Added weapon documentation

### PR DESCRIPTION
**Pull request by Mailaender**
Closes https://github.com/OpenRA/OpenRA/issues/9880.

<s>Note: I need help at the parts where the code is marked with `TODO`.</s>

### CODE REVIEW

File: OpenRA.Mods.Common/Util.cs [hunk 1]
```diff
@@ -155,5 +158,70 @@ public static IEnumerable<CPos> RandomWalk(CPos p, MersenneTwister r)
...
 			}
 		}
+
+		public static string FriendlyTypeName(Type t)
+		{
+			if (t.IsGenericType && t.GetGenericTypeDefinition() == typeof(HashSet<>))
+				return "Set of {0}".F(t.GetGenericArguments().Select(FriendlyTypeName).ToArray());
+
+			if (t.IsGenericType && t.GetGenericTypeDefinition() == typeof(Dictionary<,>))
+				return "Dictionary<{0},{1}>".F(t.GetGenericArguments().Select(FriendlyTypeName).ToArray());
```
Comment (pchote) [inline OpenRA.Mods.Common/Util.cs]: If you're going to use friendly names (which I think are silly) then this and Cached need to match.

Comment (Mailaender) [inline OpenRA.Mods.Common/Util.cs]: What is "this and Cached"?

Comment (pchote) [inline OpenRA.Mods.Common/Util.cs]: The Dictionary and Cached types.  Right now you have "2D Integer"s and "List of foo"s, but then "Dictionary<foo, bar>" which is completely inconsistent.

File: OpenRA.Mods.Common/UtilityCommands/ExtractWeaponDocsCommand.cs [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+		}
+
+		[Desc("Generate weaponry documentation in MarkDown format.")]
+		void IUtilityCommand.Run(Utility utility, string[] args)
+		{
+			// HACK: The engine code assumes that Game.modData is set.
+			Game.ModData = utility.ModData;
+
+			Console.WriteLine(
+				"This documentation is aimed at modders. It displays a template for weapon defintions " +
```
Comment (GraionDilach) [inline OpenRA.Mods.Common/UtilityCommands/ExtractWeaponDocsCommand.cs]: `defintions`

File: OpenRA.Mods.Common/Util.cs [hunk 3]
```diff
@@ -155,5 +158,70 @@ public static IEnumerable<CPos> RandomWalk(CPos p, MersenneTwister r)
...
+
+			if (t.IsGenericType && t.GetGenericTypeDefinition() == typeof(IList<>)) // TODO: why does this not work?
+				return "List of {0}".F(FriendlyTypeName(t.GetElementType()));
+
+			if (t.IsSubclassOf(typeof(Array)))
+				return "Multiple {0}".F(FriendlyTypeName(t.GetElementType()));
```
Comment (pchote) [inline OpenRA.Mods.Common/Util.cs]: Why do the documented friendly types differentiate between lists and arrays?  They are both collections of multiple things.

File: OpenRA.Mods.Common/Util.cs [hunk 4]
```diff
@@ -155,5 +158,70 @@ public static IEnumerable<CPos> RandomWalk(CPos p, MersenneTwister r)
...
+				return "3D World Position";
+
+			if (t == typeof(WDist))
+				return "1D World Distance";
+
+			if (t == typeof(WVec))
+				return "3D World Vector";
+
+			if (t == typeof(HSLColor))
+				return "hue, saturation, lightness Color";
```
Comment (pchote) [inline OpenRA.Mods.Common/Util.cs]: This also accepts hexadecimal colors, and that's how we define all of our HSLColors in our default mod yaml.

File: OpenRA.Mods.Common/Util.cs (hunk 1 above)
Comment (Mailaender) [inline OpenRA.Mods.Common/Util.cs]: Updated.

File: OpenRA.Mods.Common/Util.cs (hunk 4 above)
Comment (Mailaender) [inline OpenRA.Mods.Common/Util.cs]: What kind of notation do you suggest instead?

File: OpenRA.Mods.Common/Util.cs [hunk 5]
```diff
@@ -155,5 +158,70 @@ public static IEnumerable<CPos> RandomWalk(CPos p, MersenneTwister r)
...
+				return "{0}, {1} pairs".F(t.GetGenericArguments().Select(FriendlyTypeName).ToArray());
+
+			if (t.IsGenericType && t.GetGenericTypeDefinition() == typeof(IList<>)) // TODO: why does this not work?
+				return "List of {0}".F(FriendlyTypeName(t.GetElementType()));
+
+			if (t.IsSubclassOf(typeof(Array)))
+				return "Multiple {0}".F(FriendlyTypeName(t.GetElementType()));
+
+			if (t.IsGenericType && t.GetGenericTypeDefinition() == typeof(Primitives.Cache<,>))
+				return "Cached<{0},{1}>".F(t.GetGenericArguments().Select(FriendlyTypeName).ToArray());
```
Comment (pchote) [inline OpenRA.Mods.Common/Util.cs]: Like I said earlier, my comments about dictionaries also apply here.

Comment (pchote) [inline OpenRA.Mods.Common/Util.cs]: ...but something is very very broken if this type is ever exposed to yaml.

File: OpenRA.Mods.Common/Util.cs (hunk 3 above)
Comment (Mailaender) [inline OpenRA.Mods.Common/Util.cs]: As I stated already: this is legacy code. @ScottNZ helped me with the generics back then as I am not so confident with them. If you can help me with better code, please make some suggestions.

File: OpenRA.Mods.Common/UtilityCommands/ExtractWeaponDocsCommand.cs (hunk 2 above)
Comment (Mailaender) [inline OpenRA.Mods.Common/UtilityCommands/ExtractWeaponDocsCommand.cs]: Fixed, thanks.

File: OpenRA.Mods.Common/Util.cs (hunk 4 above)
Comment (pchote) [inline OpenRA.Mods.Common/Util.cs]: It should be combined with the `Color` case and say something like `Color (RRGGBB[AA] notation)`.

File: OpenRA.Mods.Common/Util.cs (hunk 3 above)
Comment (pchote) [inline OpenRA.Mods.Common/Util.cs]: The "list" case is new in this PR.

File: OpenRA.Mods.Common/Util.cs (hunk 5 above)
Comment (Mailaender) [inline OpenRA.Mods.Common/Util.cs]: I tracked this down to https://github.com/OpenRA/OpenRA/commit/7d0f4a261f3f2260fd05fdffebb88080683e768e and it is exposed at https://github.com/OpenRA/OpenRA/wiki/Traits#mobile but it doesn't seem to look broken in the trait documentation so I am not sure if this even fires (anymore).

File: OpenRA.Mods.Common/Util.cs (hunk 3 above)
Comment (Mailaender) [inline OpenRA.Mods.Common/Util.cs]: Yes, and the list case doesn't work for some reason so I am unable to bring them together. You are always jumping ahead making it very hard to follow what you want to say.

Comment (pchote) [inline OpenRA.Mods.Common/Util.cs]: You should be able to do something like `typeof(IEnumerable).IsAssignableFrom(t)`.

File: OpenRA.Mods.Common/Util.cs (hunk 5 above)
Comment (pchote) [inline OpenRA.Mods.Common/Util.cs]: Best to just remove it.

File: OpenRA.Mods.Common/Util.cs (hunk 3 above)
Comment (Mailaender) [inline OpenRA.Mods.Common/Util.cs]: It is a bit more complicated than that, but I came up with something after you pointed me in the right direction.

File: OpenRA.Mods.Common/Util.cs [hunk 6]
```diff
@@ -155,5 +158,64 @@ public static IEnumerable<CPos> RandomWalk(CPos p, MersenneTwister r)
...
+				return "3D World Vector";
+
+			if (t == typeof(HSLColor) || t == typeof(Color))
+				return "Color (RRGGBB[AA] notation)";
+
+			if (t == typeof(IProjectileInfo))
+				return "Projectile";
+
+			if (t == typeof(IWarhead))
+				return "Warheads";
```
Comment (pchote) [inline OpenRA.Mods.Common/Util.cs]: "Warhead" (no 's').

Comment (Mailaender) [inline OpenRA.Mods.Common/Util.cs]: Fixed.

File: OpenRA.Mods.Common/Util.cs [hunk 7]
```diff
@@ -155,5 +158,64 @@ public static IEnumerable<CPos> RandomWalk(CPos p, MersenneTwister r)
...
 			}
 		}
+
+		public static string FriendlyTypeName(Type t)
+		{
+			if (t.IsGenericType && t.GetGenericTypeDefinition() == typeof(HashSet<>))
+				return "Set of {0}".F(t.GetGenericArguments().Select(FriendlyTypeName).ToArray());
+
+			if (t.IsGenericType && t.GetGenericTypeDefinition() == typeof(Dictionary<,>))
+				return "{0}, {1} pairs".F(t.GetGenericArguments().Select(FriendlyTypeName).ToArray());
```
Comment (pchote) [inline OpenRA.Mods.Common/Util.cs]: You should be more specific here: (X,Y) pairs implies that duplicate keys with different values would be acceptable.

Comment (pchote) [inline OpenRA.Mods.Common/Util.cs]: "Mapping of {0} to {1}" would be more accurate.

Comment (Mailaender) [inline OpenRA.Mods.Common/Util.cs]: Which notation do you suggest?

File: OpenRA.Mods.Common/Util.cs [hunk 8]
```diff
@@ -155,5 +158,64 @@ public static IEnumerable<CPos> RandomWalk(CPos p, MersenneTwister r)
...
+
+			if (t.IsSubclassOf(typeof(Array)))
+				return "Multiple {0}".F(FriendlyTypeName(t.GetElementType()));
+
+			if (t.IsGenericType && t.GetGenericTypeDefinition().GetInterfaces().Any(e => e.IsGenericType && e.GetGenericTypeDefinition() == typeof(IEnumerable<>)))
+				return "Multiple {0}".F(FriendlyTypeName(t.GetGenericArguments().First()));
```
Comment (pchote) [inline OpenRA.Mods.Common/Util.cs]: "Collection of {0}" would be better here: it agrees better with "Set of {0}" above.  You can also combine this with the array check above.

File: OpenRA.Mods.Common/Util.cs (hunk 7 above)
Comment (Mailaender) [inline OpenRA.Mods.Common/Util.cs]: Updated.

File: OpenRA.Mods.Common/UtilityCommands/ExtractWeaponDocsCommand.cs [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+
+		[Desc("Generate weaponry documentation in MarkDown format.")]
+		void IUtilityCommand.Run(Utility utility, string[] args)
+		{
+			// HACK: The engine code assumes that Game.modData is set.
+			Game.ModData = utility.ModData;
+
+			Console.WriteLine(
+				"This documentation is aimed at modders. It displays a template for weapon definitions " +
+				"as well as it's contained types (warheads and projectiles) with default values and developer commentary. " +
```
Comment (pchote) [inline OpenRA.Mods.Common/UtilityCommands/ExtractWeaponDocsCommand.cs]: No apostrophe in "its" here.

File: OpenRA.Mods.Common/UtilityCommands/ExtractWeaponDocsCommand.cs [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+				"Please do not edit it directly, but add new `[Desc(\"String\")]` tags to the source code. This file has been " +
+				"automatically generated for version {0} of OpenRA.", utility.ModData.Manifest.Metadata.Version);
+			Console.WriteLine();
+
+			var toc = new StringBuilder();
+			var doc = new StringBuilder();
+			var currentNamespace = "";
+
+			var weaponInfo = new[] { typeof(WeaponInfo) }; // TODO: Does not work well with ObjectCreator so we get no default values.
```
Comment (pchote) [inline OpenRA.Mods.Common/UtilityCommands/ExtractWeaponDocsCommand.cs]: ObjectCreator handles this just fine in the normal case, so I think you're calling something incorrectly.

**Comment by pchote** [comment]
The generated page doesn't make any distinction between the base weapon object, projectiles, and warheads.  That makes it hard to understand, even for someone like me.

**Comment by pchote** [comment]
Please add some extra descriptions to the page that explains how there is a base Weapon that has exactly one Projectile and between 1 and N Warheads and then separate the projectiles and warheads into their own sections.

File: OpenRA.Mods.Common/UtilityCommands/ExtractWeaponDocsCommand.cs (hunk 10 above)
Comment (Mailaender) [inline OpenRA.Mods.Common/UtilityCommands/ExtractWeaponDocsCommand.cs]: I couldn't really find where in the code ObjectCreator actually directly worked together with WeaponInfo so I thought it just wasn't supported.

**Comment by reaperrr** [comment]
There has been no progress in nearly 3 months, so I'm closing this for now.

**Comment by Mailaender** [comment]
Rebased and resolved all problems now. Preview: https://github.com/OpenRA/OpenRA/wiki/Weapons

File: OpenRA.Mods.Common/UtilityCommands/ExtractWeaponDocsCommand.cs [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+				"as well as its contained types (warheads and projectiles) with default values and developer commentary. " +
+				"Please do not edit it directly, but add new `[Desc(\"String\")]` tags to the source code. This file has been " +
+				"automatically generated for version {0} of OpenRA.", utility.ModData.Manifest.Metadata.Version);
+			Console.WriteLine();
+
+			var toc = new StringBuilder();
+			var doc = new StringBuilder();
+			var currentNamespace = "";
+
+			var weaponInfo = Game.ModData.ObjectCreator.GetTypesImplementing<WeaponInfo>(); // TODO: Does not work well with ObjectCreator so we get no default values.
```
Comment (pchote) [inline OpenRA.Mods.Common/UtilityCommands/ExtractWeaponDocsCommand.cs]: This TODO can go away now?

Comment (Mailaender) [inline OpenRA.Mods.Common/UtilityCommands/ExtractWeaponDocsCommand.cs]: Yes, I forgot to save after deleting it.

File: OpenRA.Mods.Common/UtilityCommands/ExtractWeaponDocsCommand.cs [hunk 12]
```diff
@@ -0,0 +1,97 @@
...
+				toc.AppendLine(" * [{0}](#{1})".F(traitName, traitName.ToLowerInvariant()));
+				var traitDescLines = t.GetCustomAttributes<DescAttribute>(false).SelectMany(d => d.Lines);
+				doc.AppendLine();
+				doc.AppendLine("### {0}".F(traitName));
+				foreach (var line in traitDescLines)
+					doc.AppendLine(line);
+
+				var infos = FieldLoader.GetTypeLoadInfo(t);
+				if (!infos.Any())
+					continue;
```
Comment (pchote) [inline OpenRA.Mods.Common/UtilityCommands/ExtractWeaponDocsCommand.cs]: Style nit: newline after this please.

File: OpenRA.Mods.Common/UtilityCommands/ExtractWeaponDocsCommand.cs [hunk 13]
```diff
@@ -0,0 +1,97 @@
...
+				"as well as its contained types (warheads and projectiles) with default values and developer commentary. " +
+				"Please do not edit it directly, but add new `[Desc(\"String\")]` tags to the source code. This file has been " +
+				"automatically generated for version {0} of OpenRA.", utility.ModData.Manifest.Metadata.Version);
+			Console.WriteLine();
+
+			var toc = new StringBuilder();
+			var doc = new StringBuilder();
+			var currentNamespace = "";
+
+			var weaponInfo = Game.ModData.ObjectCreator.GetTypesImplementing<WeaponInfo>();
```
Comment (pchote) [inline OpenRA.Mods.Common/UtilityCommands/ExtractWeaponDocsCommand.cs]: Style nit: you should use `utility.ModData` instead of `Game.ModData`, and could pull `utility.ModData.ObjectCreator` to its own temporary variable.

File: OpenRA.Mods.Common/UtilityCommands/ExtractWeaponDocsCommand.cs [hunk 14]
```diff
@@ -0,0 +1,97 @@
...
+			var warheads = Game.ModData.ObjectCreator.GetTypesImplementing<IWarhead>().OrderBy(t => t.Namespace);
+			var projectiles = Game.ModData.ObjectCreator.GetTypesImplementing<IProjectileInfo>().OrderBy(t => t.Namespace);
+			var weaponryDefinitions = Enumerable.Concat(weaponInfo, Enumerable.Concat(projectiles, warheads));
+			foreach (var t in weaponryDefinitions)
```
Comment (pchote) [inline OpenRA.Mods.Common/UtilityCommands/ExtractWeaponDocsCommand.cs]: "weaponry"?

**Comment by pchote** [comment]
Adding to the milestone because this will be a big boost for the mod SDK.

File: OpenRA.Mods.Common/UtilityCommands/ExtractWeaponDocsCommand.cs (hunk 12 above)
Comment (Mailaender) [inline OpenRA.Mods.Common/UtilityCommands/ExtractWeaponDocsCommand.cs]: Fixed.

File: OpenRA.Mods.Common/UtilityCommands/ExtractWeaponDocsCommand.cs [hunk 15]
```diff
@@ -0,0 +1,99 @@
...
+				"Please do not edit it directly, but add new `[Desc(\"String\")]` tags to the source code. This file has been " +
+				"automatically generated for version {0} of OpenRA.", utility.ModData.Manifest.Metadata.Version);
+			Console.WriteLine();
+
+			var toc = new StringBuilder();
+			var doc = new StringBuilder();
+			var currentNamespace = "";
+
+			var objectCreator = utility.ModData.ObjectCreator;
+			var weaponInfo =objectCreator.GetTypesImplementing<WeaponInfo>();
```
Comment (pchote) [inline OpenRA.Mods.Common/UtilityCommands/ExtractWeaponDocsCommand.cs]: Missing space after the `=`

**Review by pchote — APPROVED** [review]
LGTM pending the style fix for this and anything else stylecop might mention

File: OpenRA.Mods.Common/UtilityCommands/ExtractWeaponDocsCommand.cs [hunk 16]
```diff
@@ -0,0 +1,99 @@
+#region Copyright & License Information
+/*
+ * Copyright 2007-2016 The OpenRA Developers (see AUTHORS)
```
Comment (obrakmann) [inline OpenRA.Mods.Common/UtilityCommands/ExtractWeaponDocsCommand.cs]: 2017 :)

**Review by obrakmann — APPROVED** [review]
👍

File: OpenRA.Mods.Common/UtilityCommands/ExtractWeaponDocsCommand.cs (hunk 16 above)
Comment (Mailaender) [inline OpenRA.Mods.Common/UtilityCommands/ExtractWeaponDocsCommand.cs]: Updated.

File: OpenRA.Mods.Common/UtilityCommands/ExtractWeaponDocsCommand.cs [hunk 17]
```diff
@@ -0,0 +1,99 @@
...
+				foreach (var info in infos)
+				{
+					var fieldDescLines = info.Field.GetCustomAttributes<DescAttribute>(true).SelectMany(d => d.Lines);
+					var fieldType = Util.FriendlyTypeName(info.Field.FieldType);
+					var defaultValue = liveTraitInfo == null ? "" : FieldSaver.SaveField(liveTraitInfo, info.Field.Name).Value.Value;
+					doc.Append("<tr><td>{0}</td><td>{1}</td><td>{2}</td>".F(info.YamlName, defaultValue, fieldType));
+					doc.Append("<td>");
+					foreach (var line in fieldDescLines)
+						doc.Append(line + " ");
+					doc.AppendLine("</td></tr>");
```
Comment (abcdefg30) [inline OpenRA.Mods.Common/UtilityCommands/ExtractWeaponDocsCommand.cs]: While we're at it, could we get a new line above this line as well? ^^

**Review by abcdefg30 — APPROVED** [review]
Looks good to me once these questions/nits are resolved. 👍 / ✅ Good work. ^^

**Comment by Mailaender** [comment]
Added some newlines to makes this more readable.

**Comment by reaperrr** [comment]
https://github.com/OpenRA/OpenRA/wiki/Changelog/_compare/765b7cdff52c8c051311c044db192a08c0894b7c...62446cdcd967d0ed1520a85108b607a2327910e6
